Show me the scatters!

cjllanwarne opened the pull request:
- [x] Depends on https://github.com/broadinstitute/wdl4s/pull/186

Shows cool scatter graphs like this!

WDL:
```
task foo {
  Int i
  command { ... }
  output {
    String str_out = i
    Int int_out = i + 1
  }
}

workflow scatter_test {
  Int x = 5
  Int y = 6
  scatter (s in [x, y]) {
    call foo { input: i = s }
  }
}
```

WomGraph
![test](https://user-images.githubusercontent.com/13006282/29892345-414410de-8d9c-11e7-889b-4e1c36d9eea0.png)

mcovarr commented:
questions: 

* Why is the `s` in the blue box of type `Any`?
* Shouldn't the `s` in the blue box connect to `i` in the white box?

mcovarr commented on src/main/scala/wdltool/graph/WomGraph.scala:
```diff
@@ -39,42 +41,102 @@ class WomGraph(graphName: String, graph: Graph) {
+  private def portLines(graphNode: GraphNode): String = graphNode match {
+    case s: ScatterNode =>
+      // Don't include the scatter expression input port here since they're added later in `internalScatterNodesAndLinks`
+      // Round up the gathered output ports so it's obvious they're being gathered..
```
.. => .

mcovarr commented on src/main/scala/wdltool/graph/WomGraph.scala:
```diff
@@ -39,42 +41,102 @@ class WomGraph(graphName: String, graph: Graph) {
+         """.stripMargin
+
+      val scatterVariableInputLink = s"$scatterVariableExpressionId -> ${scatter.scatterVariableMapping.graphInputNode.singleOutputPort.graphId} [style=dashed ltail=$scatterVariableClusterName arrowhead=empty]"
+      val outputLinks = scatter.outputMapping map { case outputPort @ ScatterGathererPort(_, _, outputToGather, _) =>
```
probably just capturing would suffice, the pattern match seems unnecessary.  might even fit on one line.  😄

mcovarr commented on src/main/scala/wdltool/graph/WomGraph.scala:
```diff
@@ -39,42 +41,102 @@ class WomGraph(graphName: String, graph: Graph) {
          |  fillcolor=${graphNode.graphFillColor};
-         |  ${portNodes.mkString(sep="\n  ")}
+         |  ${portLines(graphNode)}
+         |  ${internals.nodes.mkString(System.lineSeparator + "  ")}
```
why are you `lineSeparator`ing here and `"\n"`ing elsewhere?

mcovarr commented on src/main/scala/wdltool/graph/WomGraph.scala:
```diff
@@ -39,42 +41,102 @@ class WomGraph(graphName: String, graph: Graph) {
 
-  private def nodesAndLinks(graphNode: GraphNode): NodesAndLinks = graphNode match {
-    case other => defaultNodesAndLinks(graphNode)
+  private def upstreamLinks(graphNode: GraphNode) = for {
```
I liked having the return type

cjllanwarne commented:
@mcovarr 
- That was a TODO in WDL4S. I fixed that with the [fold-style](https://github.com/broadinstitute/wdl4s/pull/186) WOM graph building.
  - EDIT: Updated the image.
- It's because it's an expression based input node, and the expression's input is "s".

kshakir commented on src/main/scala/wdltool/graph/package.scala:
```diff
@@ -15,22 +15,19 @@ package object graph {
     })
 
-    def graphId: String = dotSafe("NODE" + graphNode.hashCode().toString)
+    def graphId: String = dotSafe("NODE" + System.identityHashCode(graphNode).toString)
```
Can you refactor this into a `def graphNodeUniqueId: Int = System.identityHashCode(graphNode)`, with a comment that this **should** be good-enough, but on some JVMs could cause unwelcome behavior. https://stackoverflow.com/questions/10645494/can-i-assume-two-objects-with-the-same-system-identityhashcode-are-the-same#10645567

mcovarr replied:
😦

kshakir commented on src/main/scala/wdltool/graph/package.scala:
```diff
@@ -40,6 +37,6 @@ package object graph {
 
     def graphName: String = dotSafe(graphNodePort.name)
-    def graphId: String = dotSafe("PORT" + graphNodePort.hashCode().toString)
+    def graphId: String = dotSafe("PORT" + System.identityHashCode(graphNodePort).toString)
```
Also, `.toString` isn't necessary.

mcovarr commented on src/main/scala/wdltool/graph/WomGraph.scala:
```diff
@@ -1,32 +1,38 @@
 import cats.implicits._
 import lenthall.validation.ErrorOr.ErrorOr
-import wdl4s.cwl.{CommandLineTool, CwlCodecs, Workflow}
+import wdl4s.cwl.CwlDecoder.decodeAllCwl
```
why importing the method and not the object?

mcovarr commented:
:+1:
